src/peft/tuners/lora/layer.py

```diff
@@ -244,8 +269,15 @@ def merge(self, safe_merge: bool = False, adapter_names: Optional[List[str]] = N
             adapter_names = self.active_adapters
 
         for active_adapter in adapter_names:
+
             if active_adapter in self.lora_A.keys():
                 base_layer = self.get_base_layer()
+                if (
```

Comment:
We have some repetition here of the same code, and it's also a bit distracting of the proper merging code. I think we can solve this by writing a `@contextmanager` that is basically

```python
@contextmanger
def some_name(base_layer):
    if (
        hasattr(base_layer, "_hf_hook")
        and isinstance(base_layer._hf_hook, AlignDevicesHook)
        and base_layer._hf_hook.offload
    ):
        base_layer._hf_hook.pre_forward(base_layer)

    yield

    if (
        hasattr(base_layer, "_hf_hook")
        and isinstance(base_layer._hf_hook, AlignDevicesHook)
        and base_layer._hf_hook.offload
    ):
        base_layer._hf_hook.weights_map = {
            name: param.to("cpu") for name, param in named_module_tensors(base_layer)
        }
        # offload weights map to disk if original device is the disk
        if torch.device("meta") in base_layer._hf_hook.original_devices.values():
            offload_folder = "offload_dir"
            offload_state_dict(offload_folder, base_layer._hf_hook.weights_map)
        base_layer._hf_hook.post_forward(base_layer, torch.tensor([]))
```

Then, here we can just use: `with some_name(base_layer) ...`. It leads to one more level of indentation, but I think that's a worthwhile trade-off.

Reply:
Good idea! Implemented for all LoRA layers